src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/SamAuthProvider.scala

```diff
@@ -0,0 +1,208 @@
+
+  //Cache lookup of pet tokens
+  val cacheExpiryTime = authConfig.getInt("cacheExpiryTime")
+  val cacheMaxSize = authConfig.getInt("cacheMaxSize")
+
+  private[leonardo] val petTokenCache = CacheBuilder.newBuilder()
+    .expireAfterWrite(cacheExpiryTime, TimeUnit.MINUTES)
+    .maximumSize(cacheMaxSize)
+    .build(
+      new CacheLoader[UserEmailAndProject, String] {
+        def load(userEmailAndProject: UserEmailAndProject) = {
+          getPetAccessTokenFromSam(userEmailAndProject.userEmail, userEmailAndProject.googleProject)
+        }
+      }
+    )
+
+  //Leo SA details -- needed to get pet keyfiles
+  private val (leoEmail, leoPem) : (WorkbenchEmail, File) = serviceAccountProvider.getLeoServiceAccountAndKey
+
+  //Given a pem, gets an access token
+  private def getAccessTokenUsingPem(email: WorkbenchEmail, pem: File): String = {
+    val credential = new GoogleCredential.Builder()
+      .setTransport(httpTransport)
+      .setJsonFactory(jsonFactory)
+      .setServiceAccountId(email.value)
+      .setServiceAccountScopes(saScopes.asJava)
+      .setServiceAccountPrivateKeyFromPemFile(pem)
+      .build()
+
+    credential.refreshToken
+    credential.getAccessToken
+  }
+
+  //Given some JSON, gets an access token
+  private def getAccessTokenUsingJson(saKey: String) : String = {
+    val keyStream = new ByteArrayInputStream(saKey.getBytes)
+    val credential = ServiceAccountCredentials.fromStream(keyStream).createScoped(saScopes.asJava)
+    credential.refreshAccessToken.getTokenValue
+  }
+
+  //"Slow" lookup of pet's access token. The cache calls this when it needs to.
+  private def getPetAccessTokenFromSam(userEmail: WorkbenchEmail, googleProject: GoogleProject): String = {
+    val samAPI = googleApi(getAccessTokenUsingPem(leoEmail, leoPem))
+    val userPetServiceAccountKey = samAPI.getUserPetServiceAccountKey(googleProject.value, userEmail.value)
+    val keyTreeMap = userPetServiceAccountKey.asInstanceOf[LinkedTreeMap[String,String]]
+    getAccessTokenUsingJson(new Gson().toJsonTree(keyTreeMap).toString)
+  }
+
+  //"Fast" lookup of pet's access token, using the cache.
+  private def getCachedPetAccessToken(userEmail: WorkbenchEmail, googleProject: GoogleProject): String = {
+    petTokenCache.get(UserEmailAndProject(userEmail, googleProject))
+  }
+
+  //A resources API if you already have a token
+  private[auth] def resourcesApi(accessToken: String): ResourcesApi = {
+    val apiClient = new ApiClient()
+    apiClient.setAccessToken(accessToken)
+    apiClient.setBasePath(authConfig.as[String]("samServer"))
+    new ResourcesApi(apiClient)
+  }
+
+  //A resources API as the given user's pet SA
+  private[auth] def resourcesApiAsPet(userEmail: WorkbenchEmail, googleProject: GoogleProject): ResourcesApi = {
+    resourcesApi(getCachedPetAccessToken(userEmail, googleProject))
+  }
+
+  private[auth] def googleApi(accessToken: String): GoogleApi = {
```

Comment:
oh, is this the sam-google API set of endpoints as opposed to the sam-resources endpoints? this probably needs better naming...